testsuite/tests/match-side-effects/check_partial.ml

```diff
@@ -58,21 +58,19 @@ let lazy_needs_partial : _ * bool t ref -> int = function
          (let
            (*match*/298 =a (field_imm 0 param/297)
             *match*/300 =o (field_mut 0 (field_imm 1 param/297)))
-           (if (isint *match*/300)
-             (if *match*/300
-               (let
-                 (*match*/303 =
-                    (let (tag/302 =a (caml_obj_tag *match*/298))
-                      (if (== tag/302 250) (field_mut 0 *match*/298)
-                        (if (|| (== tag/302 246) (== tag/302 244))
-                          (apply (field_imm 1 (global CamlinternalLazy!))
-                            (opaque *match*/298))
-                          *match*/298)))
-                  *match*/305 =o (field_mut 0 (field_imm 1 param/297)))
-                 (if (isint *match*/305) (if *match*/305 12 (exit 3))
-                   (exit 3)))
-               0)
-             (exit 3)))
+           (switch* *match*/300
+            case int 0: 0
+            case int 1:
+             (let
+               (*match*/303 =
+                  (let (tag/302 =a (caml_obj_tag *match*/298))
+                    (if (== tag/302 250) (field_mut 0 *match*/298)
+                      (if (|| (== tag/302 246) (== tag/302 244))
+                        (apply (field_imm 1 (global CamlinternalLazy!))
+                          (opaque *match*/298))
+                        *match*/298)))
+                *match*/305 =o (field_mut 0 (field_imm 1 param/297)))
+               (if (isint *match*/305) (if *match*/305 12 (exit 3)) (exit 3)))))
```

Comment:
I did not understand why this test case changed -- there was a Match_failure case before and there is still one after, so what did the temporality change in the code? This is much easier to see in the `-drawlambda` output for this test:

```diff
(setglobal Test!
  (let
    (lazy_needs_partial/274 =
       (function param/276 : int
         (catch
           (let (*match*/279 =a (field_imm 0 param/276))
             (catch
               (let
                 (*match*/280 =a (field_imm 1 param/276)
                  *match*/281 =o (field_mut 0 *match*/280))
-                 (if (isint *match*/281) (if *match*/281 (exit 2) 0)
-                   (exit 1)))
+                 (switch* *match*/281 case int 0: 0
+                                      case int 1: (exit 2)))
               with (2)
               (let
                 (*match*/288 =
                    (let
                      (lzarg/282 = *match*/279
                       tag/283 =a (caml_obj_tag lzarg/282))
                      (if (== tag/283 250) (field_mut 0 lzarg/282)
                        (if (|| (== tag/283 246) (== tag/283 244))
                          (apply (field_imm 1 (global CamlinternalLazy!))
                            (opaque lzarg/282))
                          lzarg/282)))
                  *match*/289 =a (field_imm 1 param/276)
                  *match*/290 =o (field_mut 0 *match*/289))
                 (if (isint *match*/290) (if *match*/290 12 (exit 1))
                   (exit 1)))))
         with (1)
           (raise
             (makeblock 0 (global Match_failure/20!) [0: "test.ml" 6 49])))))
    (makeblock 0 lazy_needs_partial/274)))
```

This is about the way that the check `(_, {contents = True})` is compiled in the first clause. The GADT type declaration is as follows:

```ocaml
type _ t =
  | Int : int -> int t
  | True : bool t
  | False : bool t
```

Before the compiled would do first an `isint` test to rule out `Int`, and then an `if` test to distinguish `True` from `False`. But with this commit, the compiler now trusts the totality information that tells it that the `Int` case is impossible, and generates a simpler check to differentiate `True` from `False`. (There is a similar check down below, after the `lazy ()` pattern, and that ones is pessimized as it should.)